### PR TITLE
Forward logs to different external system(s)

### PR DESCRIPTION
**Pull request by mburke5678**
https://issues.redhat.com/browse/LOG-561
https://issues.redhat.com/browse/OSDOCS-1074
Alan note; https://github.com/alanconway/cluster-logging-operator/blob/clf-url-parsing/docs/output_url_field.md#L1

### CODE REVIEW

**Comment by mburke5678** [comment]
Logging team review at: https://docs.google.com/document/d/1ORMlZH_5RPRz1G0kFlYsFFT3ANmHtcj7JTz8BFV4R-w/edit?ts=5f32f370#

File: modules/cluster-logging-collector-log-forward-configure.adoc [hunk 1]
```diff
@@ -17,60 +17,56 @@ To configure log forwarding using the API:
...
 spec:
-  disableDefaultForwarding: true <3>
-  outputs: <4>
+  outputs: <3>
    - name: elasticsearch
      type: "elasticsearch"
-     endpoint: elasticsearch.openshift-logging.svc:9200
+     url: https://elasticsearch.openshift-logging.svc:9200
      secret:
         name: fluentd
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forward-configure.adoc:31]: @mburke5678 could you please remove this output? In clusterlogforwarder, we have a default output named as `default`, if user want to set forward logs to default ES, just need to add `- default` to the `outputRefs`, no need to  define an output. You can find the details in https://issues.redhat.com/browse/LOG-706.

**Review by QiaolingTang — CHANGES_REQUESTED** [review]

File: modules/cluster-logging-collector-log-forward-es.adoc [hunk 2]
```diff
@@ -0,0 +1,104 @@
+// Module included in the following assemblies:
+//
+// * logging/cluster-logging-external.adoc
+
+[id="cluster-logging-collector-log-forward-es_{context}"]
+= Forwarding logs to an external Elasticsearch instance
+
+By default, the Cluster Logging Operator forwards all logs to the internal {product-title} Elasticsearch instance. However, you can optionally forward logs to an external Elasticsearch v5.x or v6.x instance in addition to or instead of the internal instance. You are responsible for configuring the external log aggregator to receive the logs from {product-title}.
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forward-es.adoc:8]: Start from 4.5, we use ES v6.x, so I think maybe we should say `you can optionally forward logs to an external Elasticsearch v6.x instance`.

File: modules/cluster-logging-collector-log-forward-es.adoc [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+If you want to forward logs to *only* the internal {product-title} Elasticsearch instance, you do not need to create a `ClusterLogForwarder` CR.
+====
+
+.Procedure
+
+. Create a `ClusterLogForwarder` CR YAML file similar to the following:
++
+[source,yaml]
+----
+apiVersion: "logging.openshift.io/v1alpha1"
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forward-es.adoc:25]: `apiVersion: logging.openshift.io/v1`

File: modules/cluster-logging-collector-legacy.adoc [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    overflow_action "exception"
+  </buffer>
+  <server>
+    host fluent-receiver.example.com
+    port 24224
+  </server>
+</store>
+----
+
+Sending logs using syslog::
```
Comment (gkarager) [inline modules/cluster-logging-collector-legacy.adoc:57]: @mburke5678  could you please update this part from line 57 to 88? 
With the recent changes syslog supports TLS and both RFC 3164 & 5424. 
For ref: https://github.com/openshift/cluster-logging-operator/blob/master/manifests/4.6/logging.openshift.io_clusterlogforwarders_crd.yaml#L152

**Review by gkarager — CHANGES_REQUESTED** [review]

File: modules/cluster-logging-collector-log-forward-es.adoc [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+////
+.Forwarding logs to an external Elasticsearch instance
+
+The following example forwards all logs to a single external Elasticsearch instance:
+
+[source,yaml]
+----
+apiVersion: "logging.openshift.io/v1alpha1"
+kind: "LogForwarding"
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forward-es.adoc:76]:
```
apiVersion: logging.openshift.io/v1
kind: ClusterLogForwarder
```

File: modules/cluster-logging-collector-log-forward-es.adoc [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+metadata:
+  name: instance
+  namespace: openshift-logging
+spec:
+  outputs:
+   - name: elasticsearch <1>
+     type: "elasticsearch" <2>
+     url: elasticsearch.example.com:9200 <3>
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forward-es.adoc:84]: `url: http://elasticsearch.example.com:9200`

File: modules/cluster-logging-collector-log-forward-es.adoc [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+metadata:
+  name: instance
+  namespace: openshift-logging
+spec:
+  outputs:
+   - name: elasticsearch <1>
+     type: "elasticsearch" <2>
+     url: elasticsearch.example.com:9200 <3> 
+  pipelines:
+   - name: container-logs <4>
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forward-es.adoc:86]: `container-logs` is not accurate, I was thinking we could use another name.

File: modules/cluster-logging-collector-log-forward-examples.adoc [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+
+[id=cluster-logging-collector-log-forward-examples-es]
+== Forwarding logs to an external Elasticsearch instance
+
+The following example forwards all logs to a single external Elasticsearch instance:
+
+[source,yaml]
+----
+apiVersion: "logging.openshift.io/v1alpha1"
+kind: "LogForwarding"
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forward-examples.adoc:18]:
```
apiVersion: logging.openshift.io/v1
kind: ClusterLogForwarder
```

File: modules/cluster-logging-collector-log-forward-examples.adoc [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+metadata:
+  name: instance
+  namespace: openshift-logging
+spec:
+  outputs:
+   - name: elasticsearch <1>
+     type: "elasticsearch" <2>
+     url: elasticsearch.example.com:9200 <3>
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forward-examples.adoc:26]: `url: http://elasticsearch.example.com:9200`

File: modules/cluster-logging-collector-log-forward-examples.adoc [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+
+[id=cluster-logging-collector-log-forward-examples-es-2]
+== Forwarding logs to the internal Elasticsearch instance and an external Elasticsearch instance
+
+The following example forwards all logs to a single external Elasticsearch instance and to the OpenShift Container Platform Elasticsearch instance:
+
+[source,yaml]
+----
+apiVersion: "logging.openshift.io/v1alpha1"
+kind: "LogForwarding"
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forward-examples.adoc:51]:
```
apiVersion: logging.openshift.io/v1
kind: ClusterLogForwarder
```

File: modules/cluster-logging-collector-log-forward-examples.adoc [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+metadata:
+  name: instance
+  namespace: openshift-logging
+spec:
+  outputs:
+   - name: elasticsearch <1>
+     type: "elasticsearch" <2>
+     url: elasticsearch.example.com:9200 <3>
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forward-examples.adoc:59]: `url: http://elasticsearch.example.com:9200`

File: modules/cluster-logging-collector-log-forward-examples.adoc [hunk 12]
```diff
@@ -0,0 +1,385 @@
...
+apiVersion: logging.openshift.io/v1
+kind: ClusterLogForwarder
+metadata:
+  name: instance
+  namespace: openshift-logging
+spec:
+  outputs:
+   - name: fluentd-server <1>
+     type: fluentdForward <2>
+     url: 'fluentdserver.example.com:24224' <3>
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forward-examples.adoc:94]: ` url: 'tcp://fluentdserver.example.com:24224'`

File: modules/cluster-logging-collector-log-forward-examples.adoc [hunk 13]
```diff
@@ -0,0 +1,385 @@
...
+apiVersion: "logging.openshift.io/v1alpha1"
+kind: "LogForwarding"
+metadata:
+  name: instance
+  namespace: openshift-logging
+spec:
+  outputs:
+   - name: elasticsearch-secure
+     type: "elasticsearch"
+     url: https://elasticsearch.sample.example.com:9200 <1>
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forward-examples.adoc:259]: It's `https`, we should have a secret here.

File: modules/cluster-logging-collector-log-forward-examples.adoc [hunk 14]
```diff
@@ -0,0 +1,385 @@
...
+   - name: elasticsearch-secret
+     type: "elasticsearch"
+     url: https://elasticsearch.secret.example.com:9200
+     secret:
+        name: es-secret <2>
+   - name: elasticsearch-insecure
+     type: "elasticsearch"
+     url: elasticsearch.insecure.example.com.svc:9200
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forward-examples.adoc:267]: `url: http://elasticsearch.insecure.example.com.svc:9200`

@mburke5678 out of curiosity, why do you give so many output here?

File: modules/cluster-logging-collector-log-forward-examples.adoc [hunk 15]
```diff
@@ -0,0 +1,385 @@
...
+     - application
+     outputRefs:
+     - elasticsearch-secret
+   - name: audit-logs
+     inputRefs:
+     - audit
+     outputRefs:
+     - elasticsearch-insecure
+----
+<1> To use server-authenticated TLS, specify a secure URL, with the `https` prefix.
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forward-examples.adoc:285]: Per my understanding, no matter what kind of authentication the forward use, if user want to enable TLS, then they must create a secret and specify it in the cluster log forwarder CR. @alanconway could you please help to confirm this?

File: modules/cluster-logging-collector-log-forward-examples.adoc [hunk 16]
```diff
@@ -0,0 +1,385 @@
...
+
+[id=cluster-logging-collector-log-forward-examples-multi]
+== Forwarding logs to multiple outputs
+
+The following example forwards each log source to a different log aggregator using different protocols.
+
+[source,yaml]
+----
+apiVersion: "logging.openshift.io/v1alpha1"
+kind: "LogForwarding"
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forward-examples.adoc:329]:
```
apiVersion: logging.openshift.io/v1
kind: ClusterLogForwarder
```

File: modules/cluster-logging-collector-log-forward-examples.adoc [hunk 17]
```diff
@@ -0,0 +1,385 @@
...
+metadata:
+  name: instance
+  namespace: openshift-logging
+spec:
+  outputs:
+   - name: elasticsearch-instance <1>
+     type: "elasticsearch"
+     url: elasticsearch.example.com:9200
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forward-examples.adoc:337]: url: https://elasticsearch.example.com:9200

File: modules/cluster-logging-collector-log-forward-fluentd.adoc [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+----
+apiVersion: logging.openshift.io/v1
+kind: ClusterLogForwarder
+metadata:
+  name: instance <1>
+spec:
+  outputs:
+   - name: fluentd-server <2>
+     type: fluentdForward <3>
+     url: 'https://fluentdserver.example.com:24224' <4>
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forward-fluentd.adoc:31]: `url: 'tls://fluentdserver.example.com:24224'`

File: modules/cluster-logging-collector-log-forward-update.adoc [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+// * logging/cluster-logging-external.adoc
+
+// Might not be needed. See https://issues.redhat.com/browse/LOG-654
+
+[id="cluster-logging-collector-log-forward-update_{context}"]
+= Updating log forwarding custom resources
+
+After updating to the Log Forwarding API GA, you will need to modify or recreate any existing Log Forwarding CRs that you have.
+
+If you edit your existing PRs, you must make the following modifications:
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forward-update.adoc:12]: `PRs` --> `CR`?

File: modules/cluster-logging-collector-log-forward-update.adoc [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+.Updated Log Forwarding CR
+[sourcce,yaml]
+----
+apiVersion: "logging.openshift.io/v1" <1>
+kind: "ClusterLogForwarder" <2>
+metadata:
+  name: instance
+  namespace: openshift-logging
+spec:
+  disableDefaultForwarding: true
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forward-update.adoc:23]: We don't have `disableDefaultForwarding` in the clusterlogforwarder.

File: modules/cluster-logging-collector-log-forward-update.adoc [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+     outputRefs:
+     - elasticsearch
+     - secureforward-offcluster
+----
+
+<1> Change the API from `apiVersion: "logging.openshift.io/v1alpha1"` to `apiVersion: logging.openshift.io/v1`.
+
+<2> Change the object kind from `kind: "LogForwarding"` to `kind: ClusterLogForwarder`.
+
+<3> Change the output parameter from `spec.outputs.endpoint` to `spec.output.url`
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forward-update.adoc:43]: Look at the above example, some fields are changed to the new API, some fields aren't, could we maintain a uniform format ?

File: modules/cluster-logging-collector-log-forwarding-about.adoc [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+* The internal {product-title} Elasticsearch instance does not provide secure storage for audit logs. We recommend you ensure that the system to which you forward audit logs is compliant with your organizational and governmental regulations and is properly secured. {product-title} cluster logging does not comply with those regulations.
+
+* You are responsible for creating and maintaining any additional configurations that external destinations might require, such as keys and secrets, service accounts, port openings, or global proxy configuration.
+
+The following example forwards the infrasturcture and audit logs to an external Elasticsearch instance, forwards application logs to a Kafka broker, and forwards application logs from the `my-apps-logs` namespace to the internal Elasticsearch instance. 
+
+.Sample log forwarding outputs and pipelines
+[source,yaml]
+----
+apiVersion: "logging.openshift.io/v1alpha1"
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forwarding-about.adoc:49]: apiVersion: "logging.openshift.io/v1"

File: modules/cluster-logging-collector-log-forwarding-about.adoc [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+   - name: my-app-logs 
+      application:
+        namespaces:
+        - my-project
+  pipelines:
+   - name: audit-logs <6>
+      inputRefs:
+      - audit
+      outputRefs:
+      - elasticsearch
```
Comment (QiaolingTang) [inline modules/cluster-logging-collector-log-forwarding-about.adoc:76]: Look at the example, it doesn't have an output named as `elasticsearch`.

File: modules/cluster-logging-collector-log-forward-configure.adoc (hunk 1 above)
Comment (mburke5678) [reply modules/cluster-logging-collector-log-forward-configure.adoc:31]: Removed this file.

File: modules/cluster-logging-collector-log-forward-es.adoc (hunk 2 above)
Comment (mburke5678) [reply modules/cluster-logging-collector-log-forward-es.adoc:8]: @sichvoge What did we decide on this?

File: modules/cluster-logging-collector-log-forward-examples.adoc (hunk 8 above)
Comment (mburke5678) [reply modules/cluster-logging-collector-log-forward-examples.adoc]: Removed this topic.

File: modules/cluster-logging-collector-log-forward-examples.adoc (hunk 9 above)
Comment (mburke5678) [reply modules/cluster-logging-collector-log-forward-examples.adoc]: Removed this topic.

File: modules/cluster-logging-collector-log-forward-examples.adoc (hunk 10 above)
Comment (mburke5678) [reply modules/cluster-logging-collector-log-forward-examples.adoc]: Removed this topic.

File: modules/cluster-logging-collector-log-forward-examples.adoc (hunk 11 above)
Comment (mburke5678) [reply modules/cluster-logging-collector-log-forward-examples.adoc]: Removed this topic.

File: modules/cluster-logging-collector-log-forward-examples.adoc (hunk 12 above)
Comment (mburke5678) [reply modules/cluster-logging-collector-log-forward-examples.adoc]: Removed this topic.

File: modules/cluster-logging-collector-log-forward-examples.adoc (hunk 14 above)
Comment (mburke5678) [reply modules/cluster-logging-collector-log-forward-examples.adoc]: Removed this topic.

File: modules/cluster-logging-collector-log-forward-examples.adoc (hunk 15 above)
Comment (mburke5678) [reply modules/cluster-logging-collector-log-forward-examples.adoc]: Removed this topic.

File: modules/cluster-logging-collector-log-forward-examples.adoc (hunk 16 above)
Comment (mburke5678) [reply modules/cluster-logging-collector-log-forward-examples.adoc]: Removed this topic.

File: modules/cluster-logging-collector-log-forward-examples.adoc (hunk 17 above)
Comment (mburke5678) [reply modules/cluster-logging-collector-log-forward-examples.adoc]: Removed this topic.

**Comment by mburke5678** [comment]
@anpingli @gkarager Can you please review this PR? I would like to have the docs ready when the team is ready to launch the new feature in the 4.5 z-stream.

cc. @sichvoge

File: modules/cluster-logging-collector-legacy.adoc [hunk 24]
```diff
@@ -0,0 +1,244 @@
...
+
+* For syslog, name the configuration file `syslog.conf` and specify parameters similar to the following within the `<store>` stanza:
++
+----
+<store>
+@type <tpye> <1>
+remote_syslog <syslog-server> <2>
+port 514 <3>
+hostname <host> <4>
+tag_key <key> <5>
```
Comment (gkarager) [inline modules/cluster-logging-collector-legacy.adoc:146]: @mburke5678  'tag_key' has been removed, there is only 'tag'.

**Review by gkarager — CHANGES_REQUESTED** [review]

File: modules/cluster-logging-log-forwarding-about.adoc [hunk 25]
```diff
@@ -7,30 +7,30 @@
...
-An _output_ is the destination for log data and a _pipeline_ defines simple routing for one source to one or more outputs. 
+An _output_ is the destination for log data and a _pipeline_ defines simple routing for one source to one or more outputs. The putput must be a server name or FQDN, not an IP Address, if the cluster-wide proxy using the CIDR annotation is enabled.
 
 An output can be either:
 
+* `default` to use the internal {product-title} Elasticsearch instance.
 * `elasticsearch` to forward logs to an external Elasticsearch v5.x cluster and/or the internal {product-title} Elasticsearch instance. 
-* `forward` to forward logs to an external log aggregation solution. This option uses the Fluentd *forward* protocols. 
+* `fluentdForward` to forward logs to an external log aggregation solution. This option uses the Fluentd *forward* protocols.
+* `syslog` to forward logs to a
```
[the remark this answers is not in the document]
Comment (alanconway) [reply modules/cluster-logging-log-forwarding-about.adoc]: Drop the *forward* for syslog. Fluentd has a protocol specifcally called 'forward', syslog does not.

File: modules/cluster-logging-collector-log-forward-examples.adoc [hunk 26]
```diff
@@ -0,0 +1,191 @@
...
+      - application
+      outputRefs:
+      - audit
+----
+<1> 
+<2> Kafka topic
+
+Forwarding logs to a syslog server  
+
+The following example foreards all logs to an external server that supports the RFC 3164 or 5424syslog protocol:
```
[the remark this answers is not in the document]
Comment (mburke5678) [reply modules/cluster-logging-collector-log-forward-examples.adoc]: Removed this topic.

File: modules/cluster-logging-log-forwarding-about.adoc (hunk 25 above)
[the remark this answers is not in the document]
Comment (mburke5678) [reply modules/cluster-logging-log-forwarding-about.adoc]: Removed this topic.

File: modules/cluster-logging-collector-legacy.adoc [hunk 27]
```diff
@@ -0,0 +1,240 @@
+// Module included in the following assemblies:
+//
+// * logging/cluster-logging-external.adoc
+
+[id="cluster-logging-collector-legacy_{context}"]
+= Forwarding logs using the legacy configuration method
+
+You can configure cluster logging to send logs to destinations outside of your {product-title} cluster instead of the default Elasticsearch log store using Fluentd or syslog by creating a configuration file and ConfigMap.
```
Comment (alanconway) [inline modules/cluster-logging-collector-legacy.adoc:8]: This section is referring to unmanaged mode right? It should say something about that.

File: modules/cluster-logging-collector-log-forwarding-about.adoc [hunk 28]
```diff
@@ -0,0 +1,122 @@
...
+[id="cluster-logging-collector-log-forwarding-about_{context}"]
+= About forwarding logs to third-party systems
+
+Forwarding cluster logs to external third-party systems requires a combination of _outputs_ and _pipelines_ specified in a `ClusterLogForwarder` custom resource (CR) to send logs to specific endpoints inside and outside of your {product-title} cluster. 
+
+An _output_ is the destination for log data and a _pipeline_ defines simple routing from one source to one or more outputs. 
+
+An output can be:
+
+* `default` to use the internal {product-title} Elasticsearch instance. You do not need to configure the default output. If you do configure a `default` output, you will receive an error message because the `default` output is reserved for the Cluster Logging Operator.
```
Comment (alanconway) [inline modules/cluster-logging-collector-log-forwarding-about.adoc:14]: 'default' is a built-in output _name_, used in the outputRefs section of a pipeline. The others are output _types_, used in user-defined outputs, which have user-chosen names. Need to separate them somehow.

**Review by alanconway — CHANGES_REQUESTED** [review]
Couple of minor comments, otherwise looks good to me.

File: modules/cluster-logging-collector-legacy.adoc (hunk 27 above)
Comment (mburke5678) [reply modules/cluster-logging-collector-legacy.adoc:8]: @alanconway As far as I know, unmanaged is not required. The process creates a new configMap and doesn't affect any existing objects. 
@jcantrill Can you confirm?

**Review by alanconway — APPROVED** [review]
Yep!

File: modules/cluster-logging-collector-legacy.adoc (hunk 27 above)
Comment (anpingli) [reply modules/cluster-logging-collector-legacy.adoc:8]: @mburke5678  Yes, unmanaged is not required

**Review by anpingli — COMMENTED** [review]
/lgtm

**Comment by mburke5678** [comment]
Preview: http://file.rdu.redhat.com/~mburke/logging-log-forward-ga/logging/cluster-logging-external.html

**Comment by mburke5678** [comment]
http://file.rdu.redhat.com/~mburke/logging-log-forward-ga/logging/cluster-logging-external.html

**Review by sheriff-rh — APPROVED** [review]
LGTM! Great work! 🎉 🌮 🥇

**Comment by mburke5678** [comment]
I am merging this PR as https://github.com/openshift/openshift-docs/pull/25229 depends on it. Any further comments can be tracked in a new PR.

**Pull request merged**